### PR TITLE
Mb/feature/info

### LINKED ISSUE #233: Think about adding a way to add more output on assertion failures
See [1] for how boost test does it or [2] for gtest.

[1]: https://www.boost.org/doc/libs/1_78_0/libs/test/doc/html/boost_test/test_output/test_tools_support_for_logging/contexts.html
[2]: https://google.github.io/googletest/advanced.html#adding-traces-to-assertions

### PR DESCRIPTION
**Pull request by Garados007**
Close #233

### CODE REVIEW

**Comment by Garados007** [comment]
## Update:

Current output for:
```igorpro
static Function Test7()
	variable i
	for(i = 0; i < 7; ++i)
		INFO("Loop: %d", n0 = i)
		INFO("Is smaller than 5: %d", n={ i < 5 })
		CHECK(i < 5)
	endfor
End

static Function Test8()
	INFO("empty testcase")
End
```

```
  Start of test "Unnamed"
  Entering test suite "Proc0"
  Entering test case "mymodule#Test7"
  0: is false. Assertion "CHECK(i < 5)" failed in Test7 (Proc0, line 55)
    ⓘ Loop: 5
    ⓘ Is smaller than 5: 0
  0: is false. Assertion "CHECK(i < 5)" failed in Test7 (Proc0, line 55)
    ⓘ Loop: 6
    ⓘ Is smaller than 5: 0
  Leaving test case "mymodule#Test7"
  Entering test case "mymodule#Test8"
  Test case "mymodule#Test8" doesn't contain at least one assertion
  Leaving test case "mymodule#Test8"
  Failed with 5 errors
  Leaving test suite "Proc0"
  Test finished with 5 errors
    ▶ Assertion "CHECK(i < 5)" failed in Test7 (Proc0, line 55)
    ▶ Assertion "CHECK(i < 5)" failed in Test7 (Proc0, line 55)
    ▶ Test case "mymodule#Test8" doesn't contain at least one assertion
  End of test "Unnamed"
```

**Comment by t-b** [comment]
@Garados007 Can I skip this one and give that to @MichaelHuth for review? Thanks.

**Comment by Garados007** [comment]
> Can I skip this one and give that to MichaelHuth for review? Thanks.

Yes. It wasn't your turn anyway. 😉

File: procedures/unit-testing-assertions.ipf [hunk 1]
```diff
@@ -30,6 +30,80 @@ Function FAIL()
...
+		if(!ParamIsDefault(s3))
+			s[stringLength] = s3
+			stringLength += 1
+		endif
+		if(!ParamIsDefault(s4))
+			s[stringLength] = s4
+			stringLength += 1
+		endif
+		Redimension/N=(stringLength) s
+	else
```
Comment (MichaelHuth) [inline procedures/unit-testing-assertions.ipf:68]: You can pull that together to an `elseif` here.

File: tests/Various.ipf [hunk 2]
```diff
@@ -11,6 +11,7 @@ static Constant MMD_COMBO_COUNT = 32
 Function TriggerWarningToHistory()
 	DoWindow/K HistoryCarbonCopy
 	NewNotebook/V=0/F=0 /N=HistoryCarbonCopy
+	INFO("This warning is intended to fail. The next testcase will search for it.")
```
Comment (MichaelHuth) [inline tests/Various.ipf:14]: It looks like the next testcase is not searching for it.
I think there is something like
```
	Notebook HistoryCarbonCopy, findText={"This warning is intended to fail. The next testcase will search for it.", 2}
	CHECK_EQUAL_VAR(1, V_flag)
```
missing in the CheckForWarningInHistory test case.

File: tests/VTTE.ipf [hunk 3]
```diff
@@ -710,6 +710,58 @@ static Function TestUTF()
...
+	Ensure(CheckUserPrintF("1.234568", "%.6f", emptyText, { 1.23456789 }))
+	Ensure(CheckUserPrintF("1.0MHz", "%.1W0PHz", emptyText, { 1e6 }))
+
+#if IgorVersion() > 8.00
+	Ensure(!CheckUserPrintF("", "%~d", emptyText, { 1 }, onlyErr=1))
+	Ensure(!CheckUserPrintF("", "%~", emptyText, emptyVars, onlyErr=1))
+#else
+	Ensure(CheckUserPrintF("~lld", "%~d", emptyText, { 1 }))
+#endif
+	Ensure(!CheckUserPrintF("", "@%%d", emptyText, { 1 }, onlyErr=1))
```
Comment (MichaelHuth) [inline tests/VTTE.ipf:761]: How about an additional test for the pending RTE awareness?
e.g.
```
	variable val = str2num("a")
	Ensure(CheckUserPrintF("", "", emptyText, emptyVars,onlyErr=1))
	err = GetRTError(1)

```

File: procedures/unit-testing-utils-strings.ipf [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+					err = 1
+					sprintf str, "Cannot load number %d, only %d are provided", numberIndex, numberLength
+					return str
+				endif
+
+				try
+					sprintf str, part2, numbers[numberIndex]; AbortOnRTE
+				catch
+					err = GetRTError(1)
+					sprintf str, "Cannot format number %d using format \"%s\"", numbers[numberIndex], part2
```
Comment (MichaelHuth) [inline procedures/unit-testing-utils-strings.ipf:147]: See comment above

**Review by MichaelHuth — CHANGES_REQUESTED** [review]
Looks like the UserPrintf routine works nicely.

I found a few things.

Also in the commit message of f551ad4e3abbc047d01748d3de530c2211ab9da8 it is "formatted"

**Comment by t-b** [comment]
- Can we have an example experiment/tests which trigger all the cases where the INFO should be ouptutted and/or reset?
- Another example in the documentation?

**Comment by Garados007** [comment]
I have implemented all suggestions.

**Review by MichaelHuth — APPROVED** [review]
Looks good to me

**Pull request merged**